pkg/provider/azure_vmss.go

```diff
@@ -1467,25 +1464,94 @@ func (ss *ScaleSet) ensureHostsInPool(ctx context.Context, service *v1.Service,
 				klog.ErrorS(err, "Failed to get vmss batch size", logFields...)
 				return err
 			}
-
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+			errChan := make(chan error, len(update))
+			pollerChannel := make(chan *runtime.Poller[armcompute.VirtualMachineScaleSetVMsClientUpdateResponse], len(update))
+			var pollerGroup sync.WaitGroup
+			pollerGroup.Add(1)
+			go func() {
+				defer pollerGroup.Done()
+				for {
+					select {
+					case poller, ok := <-pollerChannel:
+						if !ok {
+							// pollerChannel is closed
+							return
+						}
+						if poller == nil {
+							continue
+						}
+						pollerGroup.Add(1)
+						go func() {
+							defer pollerGroup.Done()
+							_, err := poller.PollUntilDone(ctx, &runtime.PollUntilDoneOptions{
+								Frequency: 10 * time.Second,
+							})
+							if err != nil {
+								klog.ErrorS(err, "Failed to update VMs for VMSS with new backendPoolID", logFields...)
+								errChan <- err
+							}
+						}()
+					case <-ctx.Done():
+						return
+					}
+				}
+			}()
 			klog.V(2).InfoS("Begin to update VMs for VMSS with new backendPoolID", logFields...)
-			grp, ctx := errgroup.WithContext(ctx)
-			grp.SetLimit(batchSize)
-			for instanceID, vm := range update {
-				instanceID := instanceID
-				vm := vm
-				grp.Go(func() error {
-					_, rerr := ss.ComputeClientFactory.GetVirtualMachineScaleSetVMClient().Update(ctx, meta.resourceGroup, meta.vmssName, instanceID, vm)
-					return rerr
-				})
+			if batchSize > 1 {
+				concurrentLock := make(chan struct{}, batchSize)
+				var requestGroup sync.WaitGroup
+				for instanceID, vm := range update {
+					instanceID := instanceID
+					vm := vm
+					concurrentLock <- struct{}{}
+					requestGroup.Add(1)
+					go func() {
+						defer func() {
+							requestGroup.Done()
+							<-concurrentLock
+						}()
+						poller, rerr := ss.ComputeClientFactory.GetVirtualMachineScaleSetVMClient().BeginUpdate(ctx, meta.resourceGroup, meta.vmssName, instanceID, vm, &armcompute.VirtualMachineScaleSetVMsClientBeginUpdateOptions{
+							IfMatch: vm.Etag,
+						})
+						if rerr != nil {
+							errChan <- rerr
+							return
+						}
+						if poller != nil {
+							pollerChannel <- poller
+						}
+					}()
+				}
+				requestGroup.Wait()
+				close(concurrentLock)
+			} else {
+				for instanceID, vm := range update {
+					instanceID := instanceID
+					vm := vm
+					poller, rerr := ss.ComputeClientFactory.GetVirtualMachineScaleSetVMClient().BeginUpdate(ctx, meta.resourceGroup, meta.vmssName, instanceID, vm, &armcompute.VirtualMachineScaleSetVMsClientBeginUpdateOptions{
+						IfMatch: vm.Etag,
+					})
+					if rerr != nil {
+						errChan <- rerr
+						continue
+					}
+					if poller != nil {
+						pollerChannel <- poller
+					}
+				}
 			}
-			rerr := grp.Wait()
-			if rerr != nil {
-				klog.ErrorS(err, "Failed to update VMs for VMSS", logFields...)
-				return rerr
+			close(pollerChannel)
+			pollerGroup.Wait()
```

Comment:
The loop will stop here, determined by pollerGroup.
It's ok since all vms should have pollers, and when all polller thread finish, pollerGroup.Wait return